src/main/scala/com.snowplowanalytics/iglu.schemaddl/migrations/Migration.scala

```diff
@@ -109,49 +134,31 @@ object Migration {
 
   /**
     * Creates model groups of schemas and returns them as OrderedSchemas
-    * @param schemas List of schema to create migration matrix
-    * @return List of IgluSchema as Ior.left if some groups in schemas contain
-    *         only one schema and no meaningful migrations can be provided
-    *         or List of OrderedSchemas as Ior.right if there are more than
-    *         one schemas in some groups and migrations can be provided
+    * @param orderedSchemasList List of OrderedSchemas to create migration matrix
+    * @return List of OrderedSchemas as Ior.left if migrations can not be provided
+    *         for them or List of OrderedSchemas as Ior.right if there are no problem
+    *         and migrations can be provided
     */
-  def buildMigrationMatrix(schemas: NonEmptyList[IgluSchema]): Ior[NonEmptyList[IgluSchema], NonEmptyList[OrderedSchemas]] = {
-    schemas
-      .groupByNem { case SelfDescribingSchema(SchemaMap(SchemaKey(v, n, _, SchemaVer.Full(model, _, _))), _) => (v, n, model) }
-      .toNel
-      .nonEmptyPartition { case (_, s) =>
-        buildMatrix(s.toList).map { case (_, _, ss) => OrderedSchemas(ss) } match {
-          case Nil => Left(s.head)
-          case matrix => Right(NonEmptyList.fromListUnsafe(matrix))
-        }
-      }.map(_.flatten)
-  }
-
-  def buildOrderedSchemas(schemas: NonEmptyList[IgluSchema]): Ior[NonEmptyList[IgluSchema], NonEmptyList[OrderedSchemas]] = {
-    schemas
-      .groupByNem { case SelfDescribingSchema(SchemaMap(SchemaKey(v, n, _, SchemaVer.Full(model, _, _))), _) => (v, n, model) }
-      .toNel
-      .nonEmptyPartition { case (_, s) =>
-          s match {
-            case NonEmptyList(head, Nil) => Left(head)
-            case l => Right(OrderedSchemas(l.sortBy(_.self.schemaKey)(Order.fromOrdering(SchemaKey.ordering))))
-          }
+  def buildMigrationMatrix(orderedSchemasList: NonEmptyList[OrderedSchemas]): Ior[NonEmptyList[OrderedSchemas], NonEmptyList[OrderedSchemas]] =
+    orderedSchemasList.nonEmptyPartition[OrderedSchemas, NonEmptyList[OrderedSchemas]] { l =>
+      val res = buildMatrix(l.schemas.toList).flatMap { case(_, _, m) =>
+        OrderedSchemas.unsafeSingleBuildWithoutReorder(m).toList
```

Comment:
So, in that case either we should change the argument type of `buildMigration` to something else instead of `OrderedSchemas` because we want to find migrations for `[1-0-1, 1-0-2, 1-0-3]` either or we can accept schema list like `[1-0-1, 1-0-2, 1-0-3]` which missing initial schemas however its remaining schemas are gapless in `OrderedSchemas`. What do you think about it ?

Reply:
I think _technically_ it shouldn't be even possible to create a migration without full list of schemas, for example (remember, this is just a hypothesis, we cannot implement it atm):

```
1-0-0: foo, bar optional, baz
1-1-0: remove bar
1-2-0: add quz
1-3-0: modify foo
```

If we get only `[1-2-0, 1-3-0]` then our final table will have 4 columns. With full list it will have 3 columns.

But speaking more practically, I think you're right and we need to introduce another entity. How about:

1. Renaming `OrderedSchemas` into `SchemaListFull` - to reflect that it has a connection to `SchemaList`
2. Introduce new `SchemaListSegment` - that has all properties of `SchemaListFull` except that it can miss initial or last schemas.

What do you think?

Reply:
I think that makes sense. I will work on it.